File: javascript/zebra-puzzle/zebra-puzzle.js
//
// This is only a SKELETON file for the 'Zebra Puzzle' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

// I made this global so I only have to compute the puzzle once.
let data = [{ color: '', nationality: '', drink: '', pet: '', hobby: ''}
      , { color: '', nationality: '', drink: '', pet: '', hobby: ''}
      , { color: '', nationality: '', drink: '', pet: '', hobby: ''}
      , { color: '', nationality: '', drink: '', pet: '', hobby: ''}
      , { color: '', nationality: '', drink: '', pet: '', hobby: ''}
    ]

export class ZebraPuzzle {
  constructor() {
    
    // Don't mind the suspicious ordering. I don't want a timeout.
    this.colors = ['ivory', 'yellow', 'blue', 'red', 'green']
    this.nationalities = ['Englishman', 'Japanese', 'Spainiard', 'Norwegian', 'Ukranian']
    this.drinks = ['water', 'tea', 'milk', 'coffee', 'orange juice']
    this.pets = ['dog', 'zebra', 'fox', 'horse', 'snail']
    this.hobbies = ['dancing', 'painting', 'reading', 'football', 'chess']

    // populate the givens
    data[0].nationality = 'Norwegian' // The Norwegian lives in the first house.
    data[1].color = 'blue' // The Norwegian lives next to the blue house
    data[2].drink = 'milk' // The middle house drinks milk
    if (!this.isFull())
    {
      if (!this.fillPuzzle([]))
        console.log('Fill failed')
    }
    
    console.log(this.toString())
  }

  undo(undoStack, stackPointer)
  {
    while (undoStack.length >= stackPointer)
    {
      let [house, key] = undoStack.pop()
      data[house][key] = ''
    }
  }

  // For debugging
  toString()
  {
    let msg = ''
    for(let i = 0; i < data.length; i++)
    {
      msg += `${i+1}: ${data[i].color}, ${data[i].nationality}, ${data[i].drink}, ${data[i].hobby}, ${data[i].pet}` + '\n'
    }
    return msg
  }

  isFull()
  {
    for(let house = 0; house < data.length; house++)
    {
      if (data[house].color === '' ||
        data[house].nationality === '' ||
        data[house].drink === '' ||
        data[house].pet === '' || 
        data[house].hobby === '')
        return false
    }
    return true
  }

  // Find the house that has a key with the given value
  find(key, value)
  {
    return data.findIndex((house) => house[key] === value )
  }

  // Populate if house[key] === value then house[otherKey] = otherValue
  populatePair(undoStack, key1, value1, key2, value2)
  {
    let index = this.find(key1, value1)
    if (index >= 0)
    {
      if (data[index][key2] === '')
      {
        undoStack.push([index, key2])
        data[index][key2] = value2
        return true
      }
      else if (data[index][key2] !== value2)
        return false
    }
    
    index = this.find(key2, value2)
    if (index >= 0)
    {
      if (data[index][key1] === '')
      {
        undoStack.push([index, key1])
        data[index][key1] = value1
        return true
      }
      else if (data[index][key1] !== value1)
        return false
    }
    return true
  }

  // Populate if house[key] === value then houseToRight[otherKey] = otherValue
  populatePairRight(undoStack, key1, value1, key2, value2)
  {
    let index1 = this.find(key1, value1)
    let index2 = this.find(key2, value2)
    // if you have both they had better match
    if (index1 >= 0 && index2 >=0 && (index1 + 1) !== index2)
      return false

    if (index1 >= 0 && index1 <= data.length - 2)
    {
      if (data[index1 + 1][key2] === '')
      {
        undoStack.push([index1 + 1, key2])
        data[index1 + 1][key2] = value2 
        return true 
      }
      else if (data[index1 + 1][key2] !== value2)
        return false 
    }

    if (index2 >= 1 && index2 < data.length)
    {
      if (data[index2 - 1][key1] === '')
      {
        undoStack.push([index2 - 1, key1])
        data[index2 - 1][key1] = value1
        return true 
      }
      else if (data[index2 - 1][key1] !== value1)
        return false 
    }
    return true
  }

  // These searches are lengthy to write so I broke this one in two and
  // just call it with both possible combinations (should be redundant, really).
  populatePairNeighbor_Step(undoStack, key1, value1, key2, value2)
  {
    let index = this.find(key1, value1)
    if (index >= 0)
    {
      let sides = []
      if (index - 1 >= 0)
        sides.push([index -1, data[index - 1][key2]])
      if (index + 1 < data.length)
        sides.push([index + 1, data[index + 1][key2]])

      if (sides.length === 1 && sides[0][1] !== value2 && sides[0][1] !== '')
        return false
      else if (sides.length === 1 && sides[0][1] === value2)
        return true
      else if (sides.length === 1 && sides[0][1] === '')
      {
        // only one and it is empty, fill it
        undoStack.push([sides[0][0], key2])
        data[sides[0][0]][key2] = value2
        return true
      }      
      else if (sides.length === 2 && sides[0][1] !== '' && sides[1][1] !== '' && sides[0][1] !== value2 && sides[1][1] !== value2)
      {
        // both sides not target
        return false
      }
      else if (sides.length === 2 && (sides[0][1] === value2 || sides[1][1] === value2))
      {
        // one side is target
        return true
      }
      else if (sides.length === 2 && sides[0][1] === '' && sides[1][1] !== '' && sides[1][1] !== value2)
      {
        // left empty, right not empty not target
        undoStack.push([sides[0][0], key2])
        data[sides[0][0]][key2] = value2
      }
      else if (sides.length === 2 && sides[0][1] !== '' && sides[1][1] === '' && sides[0][1] !== value2)
      {
        // right empty, let not empty not target
        undoStack.push([sides[1][0], key2])
        data[sides[1][0]][key2] = value2
      }
      
    }
    return true
  }

  // Populate if house[key] === value then houseToLeftOrRight[otherKey] = otherValue
  populatePairNeighbor(undoStack, key1, value1, key2, value2)
  {
    let index1 = this.find(key1, value1)
    let index2 = this.find(key2, value2)
    // Both populated but more than one spot apart is an error
    if (index1 >= 0 && index2 >= 0)
    {
      if ((data[index1][key1] === value1 && data[index2][key2] === value2) ||
          (data[index2][key1] === value1 && data[index1][key2] === value2))
        return true 

      if (Math.abs(index2 - index1) > 1)
        return false
    }

    if (!this.populatePairNeighbor_Step(undoStack, key1, value1, key2, value2))
      return false 
    if (!this.populatePairNeighbor_Step(undoStack, key2, value2, key1, value1))
      return false 
    return true
  }

  // Run through all the rules (except the givens -- can be filled in without futher data)
  // if a rule fails the tests return false, if no errors found return true.
  populate(undoStack)
  {
    // 1 There are five houses
    if (data.length !== 5)
      return false    
    // 2 Englishman lives in the red house
    if (!this.populatePair(undoStack, 'nationality', 'Englishman', 'color', 'red'))
      return false
    // 3 Spaniard owns a dog
    if (!this.populatePair(undoStack, 'nationality', 'Spainiard', 'pet', 'dog'))      
      return false
    // 4 Green house drinks coffee
    if (!this.populatePair(undoStack, 'color', 'green', 'drink', 'coffee'))
      return false
    // 5 Ukranian drinks tea
    if (!this.populatePair(undoStack, 'nationality', 'Ukranian', 'drink', 'tea'))      
      return false
    // 6 - Green house right of Ivory house
    //if (!this.populatePairRight(undoStack, 'color', 'green', 'color', 'ivory'))
    if (!this.populatePairRight(undoStack, 'color', 'ivory', 'color', 'green'))
      return false
    // 7 Snail owner goes dancing
    if (!this.populatePair(undoStack, 'pet', 'snail', 'hobby', 'dancing'))      
      return false
    // 8 Yellow house likes painting
    if (!this.populatePair(undoStack, 'color', 'yellow', 'hobby', 'painting'))      
      return false
    // 9 Middle house drinks milk is a given
    // 10 1st house is Norwegian is a given
    // 11 - Reading next to fox
    if (!this.populatePairNeighbor(undoStack, 'hobby', 'reading', 'pet', 'fox'))
      return false
    // 12 - Painter next to horse
    if (!this.populatePairNeighbor(undoStack, 'hobby', 'painting', 'pet', 'horse'))
      return false
    // 13 Football drinks orange juice
    if (!this.populatePair(undoStack, 'hobby', 'football', 'drink', 'orange juice'))      
      return false
    // 14 Japanese plays chess
    if (!this.populatePair(undoStack, 'nationality', 'Japanese', 'hobby', 'chess'))      
      return false
    // 15 is also a given Norwegian (1st house) is next to blue house where 2 is only neighbor 
    
    return true 
  }

  // Populate the puzzle
  fillPuzzle(undoStack)
  {
    for(let house = 0; house < data.length;house++)
    {
      for(let pair of [
        ['color', this.colors]
        , ['nationality', this.nationalities]
        , ['drink', this.drinks]
        , ['pet', this.pets]
        , ['hobby', this.hobbies]])
      {
        if (data[house][pair[0]] === '')
        {
          let key = pair[0]
          let values = pair[1]
          let used = data.filter((value) => value[key] !== '').map((item) => item[key])
          let free = values.filter((value) => !used.includes(value))
          if (free.length > 0)
          {
            for(let attempt of free)
            {
              data[house][key] = attempt
              undoStack.push([house, key])
              let stackPointer = undoStack.length
              let success = this.populate(undoStack)
              if (!success) 
                this.undo(undoStack, stackPointer)
              else 
              {
                success = this.fillPuzzle(undoStack)
                if (!success)
                  this.undo(undoStack, stackPointer)
                else
                  break
              }
            }
          }
        }        
      }
      if (data[house].color === '' || 
        data[house].nationality === '' || 
        data[house].drink === '' ||
        data[house].hobby === '' ||
        data[house].pet === '')
        return false
      if (this.isFull())
        return true
    }
    return this.populate(undoStack) && this.isFull()
  }

  waterDrinker() {
   return data[this.find('drink', 'water')].nationality
  }

  zebraOwner() {
    return data[this.find('pet', 'zebra')].nationality
  }
}

File: python/zebra-puzzle/zebra_puzzle.py
"""Zebra Puzzle - Fill in data from given constraints. 
This is a Constraint Satisfaction Problem.
"""

# I made this global so I only have to compute the puzzle once.
data = [{ 'color': '', 'nationality': '', 'drink': '', 'pet': '', 'hobby': ''}
    , { 'color': '', 'nationality': '', 'drink': '', 'pet': '', 'hobby': ''}
    , { 'color': '', 'nationality': '', 'drink': '', 'pet': '', 'hobby': ''}
    , { 'color': '', 'nationality': '', 'drink': '', 'pet': '', 'hobby': ''}
    , { 'color': '', 'nationality': '', 'drink': '', 'pet': '', 'hobby': ''}
]

colors = ['ivory', 'yellow', 'blue', 'red', 'green']
nationalities = ['Englishman', 'Japanese', 'Spainiard', 'Norwegian', 'Ukranian']
drinks = ['water', 'tea', 'milk', 'coffee', 'orange juice']
pets = ['dog', 'zebra', 'fox', 'horse', 'snail']
hobbies = ['dancing', 'painting', 'reading', 'football', 'chess']

def init():
    """
    Compute the values for the zebra puzzle
    So that the methods can query against the data.
    """
    # populate the givens
    data[0]['nationality'] = 'Norwegian' # The Norwegian lives in the first house.
    data[1]['color'] = 'blue' # The Norwegian lives next to the blue house
    data[2]['drink'] = 'milk' # The middle house drinks milk
    if not is_full():
        if not fill_puzzle([]):
            print('Fill failed')
    
def fill_puzzle(undo_stack):
    """
    Populate the puzzle
    undo_stack - Moves taken in LIFO order.
    """
    for house in range(len(data)):
        pairs = {'color': colors, 'nationality': nationalities, 'drink': drinks, 'pet': pets, 'hobby': hobbies}
        for key, values in pairs.items():
            if data[house][key] == '':
                used = [item[key] for item in data if item[key] != '']
                free = [value for value in values if value not in used]
                if len(free) > 0:
                    for attempt in free:
                        data[house][key] = attempt
                        undo_stack.append([house, key])
                        stack_pointer = len(undo_stack)
                        success = populate(undo_stack)
                        if not success:
                            undo(undo_stack, stack_pointer)
                        else:
                            success = fill_puzzle(undo_stack)
                            if not success:
                                undo(undo_stack, stack_pointer)
                            else:
                                break
        if data[house]['color'] == '' or \
        data[house]['nationality'] == '' or \
        data[house]['drink'] == '' or \
        data[house]['hobby'] == '' or \
        data[house]['pet'] == '':
            return False
      
        if is_full():
            return True
    
    return populate(undo_stack) and is_full()

def find(key, value):
    """
    Find the house that has a key with the given value
    key - the property to search
    value - the desired value of that property
    Returns -1 if not found otherwise zero based index of house with match
    """
    ret = [index for index, row in enumerate(data) if row[key] == value]
    if len(ret) == 0:
        return -1
    return ret[0]

def is_full():
    """
    Is the board filled in
    Returns True/False
    """
    return not any(data[house]['color'] == '' or \
                    data[house]['nationality'] == '' or \
                    data[house]['drink'] == '' or \
                    data[house]['pet'] == '' or \
                    data[house]['hobby'] == '' for house in range(len(data)))

def populate(undo_stack):
    """
    Run through all the rules (except the givens -- can be filled in without futher data)
    if a rule fails the tests return false, if no errors found return true.
    undo_stack - keep track of moves taken in case we need to undo a wrong series of guesses.
    """
    # 1 There are five houses
    if len(data) != 5:
        return False    
    
    # 2 Englishman lives in the red house
    if not populate_pair(undo_stack, 'nationality', 'Englishman', 'color', 'red'):
        return False
    
    # 3 Spaniard owns a dog
    if not populate_pair(undo_stack, 'nationality', 'Spainiard', 'pet', 'dog'):
        return False
    
    # 4 Green house drinks coffee
    if not populate_pair(undo_stack, 'color', 'green', 'drink', 'coffee'):
        return False
    
    # 5 Ukranian drinks tea
    if not populate_pair(undo_stack, 'nationality', 'Ukranian', 'drink', 'tea'):
        return False
    
    # 6 - Green house right of Ivory house
    if not populate_pair_right(undo_stack, 'color', 'ivory', 'color', 'green'):
        return False
    
    # 7 Snail owner goes dancing
    if not populate_pair(undo_stack, 'pet', 'snail', 'hobby', 'dancing'):
        return False
    
    # 8 Yellow house likes painting
    if not populate_pair(undo_stack, 'color', 'yellow', 'hobby', 'painting'):
        return False
    
    # 9 Middle house drinks milk is a given
    # 10 1st house is Norwegian is a given
    # 11 - Reading next to fox
    if not populate_pair_neighbor(undo_stack, 'hobby', 'reading', 'pet', 'fox'):
        return False
    
    # 12 - Painter next to horse
    if not populate_pair_neighbor(undo_stack, 'hobby', 'painting', 'pet', 'horse'):
        return False
    
    # 13 Football drinks orange juice
    if not populate_pair(undo_stack, 'hobby', 'football', 'drink', 'orange juice'):
        return False
    
    # 14 Japanese plays chess
    if not populate_pair(undo_stack, 'nationality', 'Japanese', 'hobby', 'chess'):
        return False
    
    # 15 is also a given Norwegian (1st house) is next to blue house where 2 is only neighbor 
    
    return True 

def populate_pair(undo_stack, key1, value1, key2, value2):
    """ 
    Populate if house[key] === value then house[otherKey] = otherValue
    undo_stack - moves taken in case we need to undo
    key1, value1 the first key, value pair
    key2, value2 the second key, value pair
    """
    index = find(key1, value1)
    if index >= 0:
        if data[index][key2] == '':
            undo_stack.append([index, key2])
            data[index][key2] = value2
            return True
        if data[index][key2] != value2:
            return False
    
    index = find(key2, value2)
    if index >= 0:
        if data[index][key1] == '':
            undo_stack.append([index, key1])
            data[index][key1] = value1
            return True
        if data[index][key1] != value1:
            return False
    return True

def populate_pair_right(undo_stack, key1, value1, key2, value2):
    """
    Populate if house[key] === value then houseToRight[otherKey] = otherValue
    undo_stack - moves taken in case we need to undo
    key1, value1 - first key, value pair to search/populate for
    key2, value2 - second key, value pair to search/populate for
    """
    index1 = find(key1, value1)
    index2 = find(key2, value2)
    # if you have both they had better match
    if index1 >= 0 and index2 >=0 and (index1 + 1) != index2:
        return False

    if 0 <= index1 <= len(data) - 2:
        if data[index1 + 1][key2] == '':
            undo_stack.append([index1 + 1, key2])
            data[index1 + 1][key2] = value2 
            return True 
        if data[index1 + 1][key2] != value2:
            return False 

    if 1 <= index2 < len(data):
        if data[index2 - 1][key1] == '':
            undo_stack.push([index2 - 1, key1])
            data[index2 - 1][key1] = value1
            return True 
        if data[index2 - 1][key1] != value1:
            return False 
    return True

def populate_pair_neighbor(undo_stack, key1, value1, key2, value2):
    """
    Populate if house[key] === value then houseToLeftOrRight[otherKey] = otherValue
    undo_stack - moves taken in case we need to undo
    key1, value1 - first key, value pair to search/populate for
    key2, value2 - second key, value pair to search/populate for
    """
    index1 = find(key1, value1)
    index2 = find(key2, value2)
    # Both populated but more than one spot apart is an error
    if index1 >= 0 and index2 >= 0:
        if (data[index1][key1] == value1 and data[index2][key2] == value2) or \
            (data[index2][key1] == value1 and data[index1][key2] == value2):
            return True 

        if abs(index2 - index1) > 1:
            return False

    if not populate_pair_neighbor_step(undo_stack, key1, value1, key2, value2):
        return False 
    if not populate_pair_neighbor_step(undo_stack, key2, value2, key1, value1):
        return False 
    return True

def populate_pair_neighbor_step(undo_stack, key1, value1, key2, value2):
    """
    These searches are lengthy to write so I broke this one in two and
    just call it with both possible combinations (should be redundant, really).
    undo_stack - moves taken in case we need to undo
    key1, value1 - first key, value pair to search/populate for
    key2, value2 - second key, value pair to search/populate for
    """
    index = find(key1, value1)
    if index >= 0:
        sides = []
        if index - 1 >= 0:
            sides.append([index -1, data[index - 1][key2]])
        if index + 1 < len(data):
            sides.append([index + 1, data[index + 1][key2]])

        if len(sides) == 1 and sides[0][1] != value2 and sides[0][1] != '':
            return False
        
        if len(sides) == 1 and sides[0][1] == value2:
            return True
        
        if len(sides) == 1 and sides[0][1] == '':
            # only one and it is empty, fill it
            undo_stack.append([sides[0][0], key2])
            data[sides[0][0]][key2] = value2
            return True
        
        if len(sides) == 2 and sides[0][1] != '' and sides[1][1] != '' and sides[0][1] != value2 and sides[1][1] != value2:
            # both sides not target
            return False
        
        if len(sides) == 2 and (sides[0][1] == value2 or sides[1][1] == value2):
            # one side is target
            return True
        
        if len(sides) == 2 and sides[0][1] == '' and sides[1][1] != '' and sides[1][1] != value2:
            # left empty, right not empty not target
            undo_stack.append([sides[0][0], key2])
            data[sides[0][0]][key2] = value2
        elif len(sides) == 2 and sides[0][1] != '' and sides[1][1] == '' and sides[0][1] != value2:
            # right empty, let not empty not target
            undo_stack.append([sides[1][0], key2])
            data[sides[1][0]][key2] = value2
    return True

def undo(undo_stack, stack_pointer):
    """
    Undo moves down to a given stack pointer
    undo_stack - the moves taken so far.
    stack_pointer - place on the stack to undo moves to.
    """
    while len(undo_stack) >= stack_pointer:
        [house, key] = undo_stack.pop(len(undo_stack) - 1)
        data[house][key] = ''

def drinks_water():
    """
    Find the nationality of the house that drinks water
    """
    init()
    return data[find('drink', 'water')]['nationality']

def owns_zebra():
    """
    Find the nationality of the house that has a pet zebra
    """
    init()
    return data[find('pet', 'zebra')]['nationality']
